packages/mui-base/src/utils/useControlled.ts

```diff
@@ -1,4 +1,77 @@
-/**
- * @ignore - internal hook.
- */
-export { default as useControlled } from '@mui/utils/useControlled';
+'use client';
+// TODO: uncomment once we enable eslint-plugin-react-compiler // eslint-disable-next-line react-compiler/react-compiler -- process.env never changes, dependency arrays are intentionally ignored
+/* eslint-disable react-hooks/rules-of-hooks, react-hooks/exhaustive-deps */
+import * as React from 'react';
+
+export interface UseControlledProps<T = unknown> {
+  /**
+   * Holds the component value when it's controlled.
+   */
+  controlled: T | undefined;
+  /**
+   * The default value when uncontrolled.
+   */
+  default: T | undefined;
+  /**
+   * The component name displayed in warnings.
+   */
+  name: string;
+  /**
+   * The name of the state variable displayed in warnings.
+   */
+  state?: string;
+}
+
+export function useControlled<T = unknown>({
+  controlled,
+  default: defaultProp,
+  name,
+  state = 'value',
+}: UseControlledProps<T>): [T, (newValue: T | ((prevValue: T) => T)) => void] {
+  // isControlled is ignored in the hook dependency lists as it should never change.
+  const { current: isControlled } = React.useRef(controlled !== undefined);
+  const [valueState, setValue] = React.useState(defaultProp);
+  const value = isControlled ? controlled : valueState;
+
+  if (process.env.NODE_ENV !== 'production') {
+    React.useEffect(() => {
+      if (isControlled !== (controlled !== undefined)) {
+        console.error(
+          [
+            `MUI: A component is changing the ${
```

Comment:
Maybe?

```suggestion
            `Base UI: A component is changing the ${
```

 or we could remove this, and hope that https://github.com/facebook/react/issues/28992#issuecomment-2104395451 will get fixed? 

```suggestion
            `A component is changing the ${
```

Reply:
It's an interesting problem. If this is imported in Material UI, users might get confused where does this "Base UI" come from. But it's equally true for all the other error messages (such as missing context providers). We might be better off removing the prefix in all cases. CC @atomiks

Reply:
It could be possible to configure the prefix name at runtime before any messages are logged

```js
import {log} from '@base-ui-components/react';

log.prefix('Base UI: ');
```

Then we could use utilities to throw errors:

```js
log.error('...'); // throws, prefixed
log.warn('...'); // dedupes, prefixed
```

Reply:
:+1: I like it. I wouldn't do this in this PR, though. I'll change the prefix to Base UI here, since for now only we use these utils, and create a follow-up one to configure it.